### ISSUE TITLE
[Form] Documentation about "compound" is insufficient / missing

### ISSUE THREAD

**Issue by Narretz** [symfony/symfony-docs#5213]
Currently, when your search for something like "Symfony form compound" this is the only thing this is showing up: http://symfony.com/doc/current/reference/forms/types/form.html#compound

For a beginner, this explains basically nothing. Would be nice if there was a more detailed documentation or a section in the cookbook.

**Comment by WouterJ** [comment]
I agree with you. I believe updating the description of this option to be more detailed is the best thing we can do (until we improve the custom form type cookbook recipe, as this option is pretty important in that topic).

**Comment by javiereguiluz** [comment]
I'd like to help documenting this option, but I don't understand it.

@webmozart if you have some time, could you please explain the purpose of this option so I can document it? Thanks!

**Comment by WouterJ** [comment]
@javiereguiluz this option determines whether the form contains child forms or not. E.g. TextType doesn't contain any childs, so compound is false. A custom SomeEntityType contains sub types, so compound is true (default value iirc). The DateType for instance can be both compound and not compound. When `widget` is `single_text`, compound is false (one field), otherwise compound is true (multiple input boxes).

Afaik, this option is only used internally and should not be configured/changed when using `->add(...)`.

**Comment by xabbuh** [comment]
> Afaik, this option is only used internally and should not be configured/changed when using `->add(...)`.

I am sure you are right. Afaik this is used, for example, to render forms properly (you need to iterate child views). However, this is already taken into account when the `FormView` instances are created.

**Comment by javiereguiluz** [comment]
@WouterJ in the current description of the `compound` option, you can read:

```
This (option) is independent of whether the form actually has children.
A form can be compound but not have any children at all (e.g. an empty
collection form).
```

Does this description contradicts what you said? (*"This option determines whether the form contains child forms or not"*)

**Comment by WouterJ** [comment]
The options tells if it has children or not. However, the description you quoted says: A collection form is compound, as it has children (each collection entry). However, if the property this field is bound to has no items in the collection, the form element will not display any children. This doesn't make the field a non-compound form (as it sure is).

**Comment by javiereguiluz** [comment]
@WouterJ OK thanks. I've updated the related description in the other PR.

**Comment by stof** [comment]
@javiereguiluz the wording proposed previously might indeed be confusing relating to collections (which are the reason why the option has been introduced btw, to allow supporting the case of compound forms without children and still considering them as compound when rendering).
A solution might be to say that a compound form *can* have children, while a non-compound form is a simple field (and so will always be a leaf in the Form tree btw).

**Comment by theyoux** [comment]
Hi, here is a try to make it more clear:
Current:
This option specifies if a form is compound. This is independent of whether the form actually has children. A form can be compound but not have any children at all (e.g. an empty collection form).

Proposal:
This option specifies if a field is compound. When set to true, the field may have children (think about ``DateType`` with option ``widget`` set to ``choice``).
When set to false, the field will act a leaf form type, with no children field (example : ``TextType``, or ``DateType`` with option ``widget`` set to ``single_text``).

**Comment by theyoux** [comment]
@javiereguiluz @stof @wouterj This issue is quite old, is anyone here still in charge of this part of the documentation?

**Comment by javiereguiluz** [comment]
@theyoux sadly we have no contributors for Symfony Forms docs (@wouterj submitted a PR a while ago to document data mappers and @HeahDude occasionally contributes some docs). That's why we have 38 reported issues about forms and just 1 pending PR.

**Comment by theyoux** [comment]
Ok, how can we improve that? My knowledge is quite limited and so, I need some feedback on topics like this one.
I liked @stof answer: I can try to improve my proposal until it seems ok to everyone in the place.
Current one, is really poor for people digging in documentation ;)

**Comment by xabbuh** [comment]
@theyoux Every pull request would be warmly welcome. :) I suggest that you create one with some initial wording and then we can improve it there.

**Comment by HeahDude** [comment]
Thank you @theyoux for helping us with this topic.

Here's how we could describe it IMHO:
```rst
The ``compound`` option determines whether a form is a parent form. When it is a parent, its view data is set by mapping the model data of its children instead of being get from the submitted data directly.

It does not mean it has children, think of a ``CollectionType`` without any children, or a ``ChoiceType`` without any checkboxes checked.
```

**Comment by vudaltsov** [comment]
@HeahDude , I really like `determines whether a form is a parent form` but I don't agree with `ChoiceType without any checkboxes checked`. ChoiceType has children when `choices` option is not empty and when `expanded=true`, no matter whether the checkboxes/radios are checked or not.

I actually think that it's not really important to emphasize the ability of compound form __to have no children__. It's insignificant for understanding and it's something that you can get/notice later.

The ability __to have children__ - that's what's important for the definition.

So the first sentence could be: `The compound option determines whether a form of this type can have children, in other words is a parent form.`

**Comment by javiereguiluz** [comment]
Fixed by #9283.

**Comment by Narretz** [comment]
Cool to see that this has been addressed. I'm currently not using Symfony but when I come back I'll appreciate the docs 😊
